File: app/src/main/java/com/example/lumuttest/data/repository/TodoRepositoryImpl.java
package com.example.lumuttest.data.repository;

import com.example.lumuttest.data.remote.ApiService;
import com.example.lumuttest.data.remote.RetrofitClient;
import com.example.lumuttest.data.remote.Todo;
import com.example.lumuttest.domain.repository.TodoRepository;

import java.io.IOException;
import java.util.List;

import retrofit2.Response;

public class TodoRepositoryImpl implements TodoRepository {
    private final ApiService apiService;

    public TodoRepositoryImpl() {
        this.apiService = RetrofitClient.getClient().create(ApiService.class);
    }

    @Override
    public List<Todo> getTodos() throws Exception {
        Response<List<Todo>> response = apiService.getTodos().execute();
        if (response.isSuccessful()) {
            return response.body();
        } else {
            throw new IOException("Error fetching todos: " + response.code());
        }
    }

    @Override
    public Todo getTodoById(int id) throws Exception {
        Response<Todo> response = apiService.getTodoById(id).execute();
        if (response.isSuccessful()) {
            return response.body();
        } else {
            throw new IOException("Error fetching todo: " + response.code());
        }
    }
}

File: app/src/main/java/com/example/lumuttest/data/repository/TodoRepositoryImpl.kt
package com.example.lumuttest.data.repository

import android.util.Log
import com.example.lumuttest.data.remote.ApiService
import com.example.lumuttest.data.remote.RetrofitClient
import com.example.lumuttest.data.remote.Todo
import com.example.lumuttest.domain.repository.TodoRepository
import java.io.IOException

class TodoRepositoryImpl : TodoRepository {
    private val apiService: ApiService = RetrofitClient.getClient().create(ApiService::class.java)

    override suspend fun getTodos(): List<Todo> {
        return try {
            val todos = apiService.getTodos()
            Log.d("TodoRepository", "getTodos response: $todos")
            todos
        } catch (e: Exception) {
            Log.e("TodoRepository", "Error fetching todos", e)
            throw IOException("Error fetching todos", e)
        }
    }

    override suspend fun getTodoById(id: Int): Todo? {
        return try {
            val todo = apiService.getTodoById(id)
            Log.d("TodoRepository", "getTodoById response: $todo")
            todo
        } catch (e: Exception) {
            Log.e("TodoRepository", "Error fetching todo", e)
            throw IOException("Error fetching todo", e)
        }
    }
}
